Source/levels/gendung.cpp

```diff
@@ -95,83 +95,42 @@ std::optional<Size> GetSizeForThemeRoom(int floor, Point origin, int minSize, in
 	if (origin.x + maxSize > DMAXX && origin.y + maxSize > DMAXY) {
 		return {}; // Original broken bounds check, avoids lower right corner
 	}
-	if (origin.x + minSize > DMAXX || origin.y + minSize > DMAXY) {
-		return {}; // Skip definit OOB cases
-	}
 	if (IsNearThemeRoom(origin)) {
 		return {};
 	}
 
-	bool yFlag = true;
-	bool xFlag = true;
-	int xCount = 0;
-	int yCount = 0;
-
-	int xArray[20] = {};
-	int yArray[20] = {};
-
-	for (int ii = 0; ii < maxSize; ii++) {
-		if (xFlag && origin.y + ii < DMAXY) {
-			for (int xx = origin.x; xx < origin.x + maxSize && xx < DMAXX; xx++) {
-				if (dungeon[xx][origin.y + ii] != floor) {
-					if (xx >= /* origin.x + */ minSize) {
-						break;
-					}
-					xFlag = false;
-				} else {
-					xCount++;
-				}
-			}
-			if (xFlag) {
-				xArray[ii] = xCount;
-				xCount = 0;
-			}
-		}
-		if (yFlag && origin.x + ii < DMAXX) {
-			for (int yy = origin.y; yy < origin.y + maxSize && yy < DMAXY; yy++) {
-				if (dungeon[origin.x + ii][yy] != floor) {
-					if (yy >= /* origin.y + */ minSize) {
-						break;
-					}
-					yFlag = false;
-				} else {
-					yCount++;
-				}
-			}
-			if (yFlag) {
-				yArray[ii] = yCount;
-				yCount = 0;
-			}
-		}
-	}
+	int maxWidth = std::min(maxSize, DMAXX - origin.x);
+	int maxHeight = std::min(maxSize, DMAXY - origin.y);
 
-	for (int ii = 0; ii < minSize; ii++) {
-		if (xArray[ii] < minSize || yArray[ii] < minSize) {
-			return {};
-		}
-	}
+	Size room { maxWidth, maxHeight };
 
-	for (int ii = 0; ii < minSize; ii++) {
-		if (xArray[ii] < minSize || yArray[ii] < minSize) {
-			return {};
+	for (int i = 0; i < maxSize; i++) {
+		int width = 0;
+		for (int x = 0; x < room.width && i < maxHeight; x++) {

```

Comment:
`maxHeight` and `i` don't change within the loop so I think this will be clearer if you wrap the loop with `if (i < maxHeight)`